pkg/server/credtemplate/builder_test.go

```diff
@@ -0,0 +1,1188 @@
+					SubjectKeyId:          publicKeyID,
+					AuthorityKeyId:        parentKeyID,
+					NotBefore:             notBefore,
+					// Downstream CAs have historically been signed using the default X509-SVID TTL
+					NotAfter: x509SVIDNotAfter,
+				}
+				if tc.overrideExpected != nil {
+					tc.overrideExpected(expected)
+				}
+				require.Equal(t, expected, template)
+			})
+		})
+	}
+}
+
+func TestBuildServerX509SVIDTemplate(t *testing.T) {
+	for _, tc := range []struct {
+		desc             string
+		overrideConfig   func(config *credtemplate.Config)
+		overrideParams   func(params *credtemplate.ServerX509SVIDParams)
+		overrideExpected func(expected *x509.Certificate)
+		expectErr        string
+	}{
+		{
+			desc: "defaults",
+		},
+		{
+			desc: "fail to get serial number",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.NewSerialNumber = failNewSerialNumber
+			},
+			expectErr: "failed to get new serial number: oh no",
+		},
+		{
+			desc: "invalid parent chain",
+			overrideParams: func(params *credtemplate.ServerX509SVIDParams) {
+				params.ParentChain = nil
+			},
+			expectErr: "parent chain required to build X509-SVID template",
+		},
+		{
+			desc: "invalid public key",
+			overrideParams: func(params *credtemplate.ServerX509SVIDParams) {
+				params.PublicKey = nil
+			},
+			expectErr: "x509: unsupported public key type: <nil>",
+		},
+		{
+			desc: "override X509SVIDTTL",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.X509SVIDTTL = credtemplate.DefaultX509SVIDTTL * 2
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.NotAfter = now.Add(credtemplate.DefaultX509SVIDTTL * 2)
+			},
+		},
+		{
+			desc: "ttl capped by parent chain",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.X509SVIDTTL = parentTTL + time.Hour
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.NotAfter = now.Add(parentTTL)
+			},
+		},
+		{
+			desc: "override X509SVIDSubject",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.X509SVIDSubject = pkix.Name{CommonName: "OVERRIDE"}
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.Subject = pkix.Name{
+					CommonName: "OVERRIDE",
+					ExtraNames: []pkix.AttributeTypeAndValue{x509svid.UniqueIDAttribute(serverID)},
+				}
+			},
+		},
+		{
+			desc: "single composer",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.CredentialComposers = []credentialcomposer.CredentialComposer{fakeCC{id: 1}}
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.Subject.CommonName = "OVERRIDE-1"
+				expected.DNSNames = []string{"OVERRIDE-1"}
+				expected.ExtraExtensions = []pkix.Extension{{Id: makeOID(1), Value: []byte{1}}}
+			},
+		},
+		{
+			desc: "two composers",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.CredentialComposers = []credentialcomposer.CredentialComposer{fakeCC{id: 2}, fakeCC{id: 2}}
```

Comment:
IDs are the same for both composers (same in all `two composers` test cases

Reply:
Yikes. Thanks.